### PR TITLE
ADEN-1834: Interstitial ads on mercury

### PR DESCRIPTION
**Pull request by nandy-andy**
We implemented interstitial ads on Oasis and we're using there `UiFactory` and its regular modals. We also wanted to have the interstitial product on Mercury and changes below allows us to have it. The changes includes:
- extracted exposed functionality `loadCustomAd()` to a separate module: `ext.wikia.adEngine.customAdsLoader`,
- exposed `loadCustomAd()` on Mercury as well (in the run file),
- new method `createAdIframe()` reused for Oasis&Mercury in `show()` funciton of `modal.js`,
- added `wikia-ad-iframe` class if an iframe is created by `iframeWriter`.

This pull request is connected to https://github.com/Wikia/mercury/pull/630.

To test if it's work open a mercury page in your browser and execute in JavaScript console following code:

``` js
window.loadCustomAd && window.loadCustomAd({
  type: 'modal',
  width: 300,
  height: 250,
  code: 'An ad placeholder'
});
```

### CODE REVIEW

File: extensions/wikia/AdEngine/js/mercury.run.js [hunk 1]
```diff
@@ -1,5 +1,14 @@
 /*global require*/
-require(['ext.wikia.adEngine.messageListener'], function (messageListener) {
+require([
+	'ext.wikia.adEngine.messageListener',
+	'ext.wikia.adEngine.customAdsLoader'
```
Comment (rogatty) [inline extensions/wikia/AdEngine/js/mercury.run.js]: Shouldn't it be kept in an alphabetic order?

Comment (gabrys) [inline extensions/wikia/AdEngine/js/mercury.run.js]: Yeap, alphabetical order is better, but not a biggie.

File: resources/wikia/modules/iframeWriter.js [hunk 2]
```diff
@@ -28,6 +28,11 @@ define('wikia.iframeWriter', [
 		iframe.scrolling = 'no';
 		iframe.width = params.width;
 		iframe.height = params.height;
+
+		if (iframe.classList) {
+			iframe.classList.add('wikia-ad-iframe');
```
Comment (nandy-andy) [inline resources/wikia/modules/iframeWriter.js]: After giving it a second thought to this I'm wondering if it shouldn't be `wikia-iframe` or if we shouldn't add possibility to pass classes in the `params` object.

Comment (gabrys) [inline resources/wikia/modules/iframeWriter.js]: +1 to adding an optional class param to iframeWriter

File: extensions/wikia/AdEngine/js/template/modal.js [hunk 3]
```diff
@@ -19,7 +21,20 @@ define('ext.wikia.adEngine.template.modal', [
 	 */
 	function show(params) {
 		log(['showNew', params], 'debug', logGroup);
```
Comment (gabrys) [inline extensions/wikia/AdEngine/js/template/modal.js]: log(['show'

File: extensions/wikia/AssetsManager/config.php [hunk 4]
```diff
@@ -73,6 +73,7 @@
 		'//extensions/wikia/AdEngine/js/AdLogicPageParams.js',
 		'//extensions/wikia/AdEngine/js/AdLogicPageViewCounter.js',
 		'//extensions/wikia/AdEngine/js/DartUrl.js',
+		'//extensions/wikia/AdEngine/js/CustomAdsLoader.js',
```
Comment (gabrys) [inline extensions/wikia/AssetsManager/config.php]: CDE ;-)

File: resources/wikia/modules/iframeWriter.js [hunk 5]
```diff
@@ -17,6 +17,7 @@ define('wikia.iframeWriter', [
 	 * @param {string} params.code - code to put into the iframe
 	 * @param {number} params.width - desired width of the iframe
 	 * @param {number} params.height - desired height of the iframe
+	 * @param {array} [params.classes] - custom classes of the iframe
```
Comment (gabrys) [inline resources/wikia/modules/iframeWriter.js]: Let's have the {string} params.class param and just assign it to iframe.class when creating iframe.

File: tests/karma/js-unit.conf.js [hunk 6]
```diff
@@ -58,6 +58,7 @@ module.exports = function (config) {
 			'resources/wikia/modules/thumbnailer.js',
 			'resources/wikia/modules/uniqueId.js',
 			'resources/wikia/modules/scrollToLink.js',
+			'resources/wikia/modules/iframeWriter.js',
```
Comment (gabrys) [inline tests/karma/js-unit.conf.js]: Let's sort them as well, seems like only scrollToLink and iframeWriter are out of order.

File: extensions/wikia/AdEngine/js/template/modal.js [hunk 7]
```diff
@@ -18,8 +20,21 @@ define('ext.wikia.adEngine.template.modal', [
...
+		var skin = adContext.getContext().targeting.skin;
 
+		if (skin === 'oasis') {
+			log(['showNew desktop modal'], 'debug', logGroup);
+			createAndShowDesktopModal(params);
+		}
+
+		if (skin === 'mercury') {
+			log(['showNew mobile (Mercury) modal'], 'debug', logGroup);
+			win.Mercury.Modules.Ads.getInstance().openLightbox(createAdIframe(params));
```
Comment (rogatty) [inline extensions/wikia/AdEngine/js/template/modal.js]: What if something breaks and there is an `undefined` in the chain? Should we add some checks just to be safe?

Comment (nandy-andy) [inline extensions/wikia/AdEngine/js/template/modal.js]: I don't think we want to have any graceful error handling here. If our tests don't catch it then somebody will sooner or later notice error in the console and we'll fix it :)

File: resources/wikia/modules/iframeWriter.js [hunk 8]
```diff
@@ -17,6 +17,7 @@ define('wikia.iframeWriter', [
 	 * @param {string} params.code - code to put into the iframe
 	 * @param {number} params.width - desired width of the iframe
 	 * @param {number} params.height - desired height of the iframe
+	 * @param {string} [params.classes] - custom class(es) of the iframe (coma-separated if more than one)
```
Comment (rogatty) [inline resources/wikia/modules/iframeWriter.js]: Why the square brackets? Are you sure they should be separated by coma and not by space?

Comment (nandy-andy) [inline resources/wikia/modules/iframeWriter.js]: 1. It's a notation we use in AdEng - square brackets = optional parameter and PhpStorm accepts it :smile_cat: 
2. Good catch! I'll correct it in a minute :cake:

File: resources/wikia/modules/spec/iframeWriter.spec.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+		expect(iframe.frameborder).toBe('no');
+		expect(iframe.scrolling).toBe('no');
+		expect(iframe.onload).toBeDefined();
+	});
+
+	it('Custom classes', function () {
+		var settings = testIframeSettings,
+			iframe;
+
+		settings.classes = 'class-one,class-two,class-three';
```
Comment (rogatty) [inline resources/wikia/modules/spec/iframeWriter.spec.js]: Should these be separated by spaces too?

Comment (nandy-andy) [inline resources/wikia/modules/spec/iframeWriter.spec.js]: Yep, what was I thinking? ;)

**Comment by rogatty** [comment]
LGTM

**Comment by nandy-andy** [comment]
The failing unit tests isn't connected to the changes at all. It says:
```
Fatal error: Class 'WikiaMobileService' not found in /var/lib/jenkins/workspace/unitTests_php_branches/extensions/wikia/WikiaMobile/tests/WikiaMobileServiceTest.php on line 17
```

File: resources/wikia/modules/iframeWriter.js (hunk 8 above)
Comment (gabrys) [inline resources/wikia/modules/iframeWriter.js]: [] means optional: 
http://usejsdoc.org/tags-param.html#optional-parameters-and-default-values

**Pull request merged**